stdlib/experimental/dynamic/dynamic.go

```diff
@@ -21,10 +27,94 @@ var dynamicConv = values.NewFunction(
 			return nil, err
 		}
 
+		// Nothing to do if the incoming value is already a dynamic
+		if v.Type().Nature() == semantic.Dynamic {
+			v.Retain()
+			return v, nil
+		}
+
 		// FIXME(onelson): do we need to special case nulls here?
 		//   Ex: if the input is null, should the output be null or a
 		//   Dynamic with a null in it? My sense is the latter.
 		return values.NewDynamic(v), nil
 	},
 	false,
 )
+
+var asArray = values.NewFunction(
+	"asArray",
+	runtime.MustLookupBuiltinType("experimental/dynamic", "asArray"),
+	func(ctx context.Context, args values.Object) (values.Value, error) {
+		arguments := interpreter.NewArguments(args)
+		v, err := arguments.GetRequired("v")
+		if err != nil {
+			return nil, err
+		}
+
+		if v.IsNull() {
+			panic("TODO")
+		}
+
+		d := v.Dynamic()
+		inner := d.Inner()
+		if inner.Type().Nature() != semantic.Array {
+			return nil, errors.Newf(codes.Invalid, "unable to convert %s to array", inner.Type())
+		}
+		arr := inner.Array()
+		elmType, err := arr.Type().ElemType()
+		if err != nil {
+			return nil, err
+		}
+
+		// The contract for this function says it accepts a dynamic and produces
+		// an array of dynamic.
+		// Note that just because we've verified the argument `v` was a dynamic
+		// value holding an array, it doesn't mean the elements inside that array
+		// are wrapped in dynamic.
+		// Therefore, check to see if the elements are dynamic and wrap them if they aren't.
+		if elmType.Nature() == semantic.Dynamic {
+			arr.Retain()
+			return arr, nil
+		} else {
+			elems := make([]values.Value, arr.Len())
+			for i := 0; i < arr.Len(); i++ {
+				v := arr.Get(i)
+				v.Retain()
+				elems[i] = values.NewDynamic(v)
+			}
+			dynArr := values.NewArrayWithBacking(semantic.NewArrayType(semantic.NewDynamicType()), elems)
+			return dynArr, nil
+		}
+	},
+	false,
+)
+
+var equal = values.NewFunction(
+	"_equal",
+	runtime.MustLookupBuiltinType("experimental/dynamic", "_equal"),
+	func(ctx context.Context, args values.Object) (values.Value, error) {
+		arguments := interpreter.NewArguments(args)
+		a, err := arguments.GetRequired("a")
+		if err != nil {
+			return nil, err
+		}
+		if a.IsNull() {
+			return values.NewBool(false), nil
```

Comment:
It seems like for testing purposes, you might want to consider nulls as equal?

Reply:
iirc in the spec we say that nulls are never equal, but you're right that it may make sense in this case :thinking:

Reply:
Okay, so here's what I have rationale-wise.

The `a.IsNull()` check here is for the "untyped null" case where the argument `a` is null and didn't come from some known-typed column value. This is to say, the argument isn't even a `dynamic`.

I would expect the outcome to match whatever I got from nulls that are properly wrapped in a `dynamic`.

```
$ go run ./cmd/flux
> import dyn "experimental/dynamic"
> import "internal/debug"
> untypedNull = dyn.dynamic(v: debug.null())
> dyn._equal(a: untypedNull, b: untypedNull)
false
> intNull = dyn.dynamic(v: debug.null(type: "int"))
> dyn._equal(a: intNull, b: intNull)
false
```

For dynamic-wrapped nulls, both typed and untyped, I see `false` where `a` and `b` are the same flavor of null.

I think this makes sense if we're talking about value equality while disregarding identity. My discussion around "equivalency" earlier during show and tell obviously doesn't hold true here since I'd argue these are indeed equivalent but not equal.

Reply:
It shouldn't be possible to pass a non-dynamic value to `_equal` because of its type signature, right?

Is `_equal` explicitly just for testing? If so, this gets an empty diff, which i think suggests `dynamic._equal` should follow suit:
```
> import "internal/debug"
> import "testing"
> testing.assertEqualValues(want: debug.null(type: "int"), got: debug.null(type: "int"))
Result: _result
>
```

Reply:
Interesting.

Yes, my understanding is static checks would have failed us if we pass values that are not dynamic and reach that particular branch in the impl. Today I can only think it would have to be some sort of programmer error on the Go side, or otherwise refactors to how we treat dynamic statically.

`_equal` is explicitly for testing, yes. I'd expect once we have the rest of the cast functions in place, we'd use those to write more regular acceptance tests that compare tables.

_If we want to say nulls are equal_ we would need to be ready to no longer rely on `Value.Equal()` method and do something else. I can patch in a naive check to see if a and b themselves are both null, but I think if the nulls are deeper down in the value we'll get the wrong answer via `Value.Equal()`.

Reply:
What about:
```js
_equal = (a, b) => display(v: a) == display(v: b)
```
Two dynamic values would be equal if their display strings are equal. We could call it something less loaded like `_isNotDistinct` to avoid confusion with `==` semantics in Flux.

As an aside, do we want to allow `==` to be used with dynamic values? It seems like maybe we want to produce an error for that, to force users to cast first?

Reply:
Yeah, I think leveraging display is a good way to go. One issue might be field ordering of records in the output but if we hit that case for some reason, fixing it in `display()` will have bigger benefit.

I'll rebase and get that going. I can even move the definition of this into the test package, rather than exposing it. Win-win.

Reply:
(actually, I'll leave it in the main package for now since it might be useful in the REPL as we hack through this stuff)